gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy

```diff
@@ -66,6 +85,104 @@ class CreateManifestTask extends DefaultTask {
+                def productId = "${recommendedDep.productGroup()}:${recommendedDep.productName()}".toString()
+                if (mavenCoordsByProductIds.containsKey(productId)) {
+                    def othercoord = mavenCoordsByProductIds.get(productId)
+                    throw new GradleException("Duplicate product dependency recommendations found for " +
+                            "'${productId}' in '${coord}' and '${othercoord}'")
+                }
+                mavenCoordsByProductIds.put(productId, coord)
+            }
+        }
+
+        Map<String, RecommendedProductDependency> recommendedDepsByProductId = [:]
+        allRecommendedProductDeps.each { recommendedDep ->
+            def productId = "${recommendedDep.productGroup()}:${recommendedDep.productName()}".toString()
+            recommendedDepsByProductId.put(productId, recommendedDep)
+        }
+
+        Set<String> seenRecommendedProductIds = []
+        def dependencies = []
+        productDependencies.each { productDependency ->
+            def productId = "${productDependency.productGroup}:${productDependency.productName}".toString()
+
+            if (!productDependency.detectConstraints) {
+                dependencies.add(jsonMapper.convertValue(productDependency, Map))
+            } else {
+                if (!recommendedDepsByProductId.containsKey(productId)) {
+                    throw new GradleException("Product dependency '${productId}' has constraint detection enabled, " +
+                            "but could not find any recommendations in ${productDependenciesConfig.name} configuration")
+                }
+                def recommendedProductDep = recommendedDepsByProductId.get(productId)
+                try {
+                    dependencies.add(jsonMapper.convertValue(
+                            new ProductDependency(
+                                    productDependency.productGroup,
+                                    productDependency.productName,
+                                    recommendedProductDep.minimumVersion(),
+                                    recommendedProductDep.maximumVersion(),
+                                    recommendedProductDep.recommendedVersion()),
+                            Map))
+
+                } catch (IllegalArgumentException e) {
+                    def mavenCoordSource = mavenCoordsByProductIds.get(productId)
+                    throw new GradleException("IllegalArgumentException encountered when generating " +
+                            "ProductDependency from recommended dependency for ${productId} from ${mavenCoordSource}",
+                            e)
+                }
+            }
+
+            seenRecommendedProductIds.add(productId)
+        }
+
+        def unseenProductIds = new HashSet<>(recommendedDepsByProductId.keySet())
+        seenRecommendedProductIds.each { unseenProductIds.remove(it) }
```

Comment:
i think you can do set subtraction in groovy: https://stackoverflow.com/questions/5713665/opposite-of-intersect-in-groovy-collections